File: Clojure/problem042.clj
#! /usr/bin/env clojure
; Problem 42 - Project Euler
; http://projecteuler.net/index.php?section=problems&id=042
(defn triangle-numbers [nmax]
  (for [i (range 1 nmax)]
    (/ (* i (inc i))
       2)))

(defn char2score [c]
  (- (int c) 64))

(defn alphabetical-score [word]
  (->> (seq word)
       (filter #(not= % \"))
       (map char2score)
       (reduce +)))

(def tns (triangle-numbers 100))
(->> (-> (slurp "data/p042_words.txt")
         (clojure.string/split #","))
     (map alphabetical-score)
     (filter #(>= (.indexOf tns %) 0))
     (count)
     (println))

File: Python/problem042.py
#! /usr/bin/env python3
"""
Problem 42 - Project Euler
http://projecteuler.net/index.php?section=problems&id=042
"""
import math
import csv
from itertools import chain

filename = "data/p042_words.txt"
alphabets = {a: i + 1 for i, a in enumerate("abcdefghijklmnopqrstuvwxyz")}


def getAlphabeticalScore(word):
    return sum(alphabets[c] for c in word)


def triangleNumbers(nmax):
    return [i * (i + 1) / 2 for i in range(1, nmax + 1)]


if __name__ == "__main__":
    csv_file = open(filename)
    f = csv.reader(csv_file)
    fin = sorted(chain.from_iterable(f))
    csv_file.close()

    ans = 0
    tn = triangleNumbers(100)
    for w in fin:
        score = getAlphabeticalScore(w.rstrip().lower())
        if score in tn:
            ans += 1

    print(ans)
